tests/chainer_tests/test_variable.py

```diff
@@ -983,358 +981,325 @@ def v():
-        assert x_var.device == expected_device
-        assert x.shape == x_var.shape
-        assert x.dtype == x_var.dtype
+        # Check x.
+        x_src = self.x_src
+        assert x.shape == x_src.shape
+        assert x.dtype == x_src.dtype
         np.testing.assert_array_equal(
-            backend.CpuDevice().send(x_var.data), backend.CpuDevice().send(x))
-
-        if set_grad_var:
-            assert isinstance(x_var.grad, cuda.cupy.ndarray)
-            assert x_var.grad.device.id == expected_device_id
-            assert x_var.grad_var.device == expected_device
-            assert gx.shape == x_var.grad.shape
-            assert gx.dtype == x_var.grad.dtype
+            _numpy_device.send(x.array), _numpy_device.send(x_src))
+
+        # Check gx.
+        gx_src = self.gx_src
+        if self.x_with_grad:
+            assert x.grad.shape == gx_src.shape
+            assert x.grad.dtype == gx_src.dtype
+            assert x.grad_var.shape == gx_src.shape
+            assert x.grad_var.dtype == gx_src.dtype
             np.testing.assert_array_equal(
-                backend.CpuDevice().send(x_var.grad),
-                backend.CpuDevice().send(gx))
-            assert x_var.grad_var is not None
-            assert x_var.grad_var.node is not None
+                _numpy_device.send(x.grad), _numpy_device.send(gx_src))
         else:
-            assert x_var.grad is None
-            assert x_var.grad_var is None
+            assert x.grad is None
+            assert x.grad_var is None
 
-        orig_device = backend.get_device_from_array(x)
-        if orig_device == expected_device:
-            assert x_var.data is x
-            assert x_var.grad is gx
-        else:
-            assert x_var.data is not x
-            assert not set_grad_var or x_var.grad is not gx
+    def test_to_device_method_allow_unchain(
+            self, src_backend_config, dst_backend_config):
+        self.setup(src_backend_config)
 
-        assert x_var.xp is not chainerx
+        self.to_device_method(dst_backend_config.device, True)
 
-    def test_to_gpu_from_cpu(self):
-        self.check_to_gpu(self.x, self.gx)
+        self.check_to_device_method(
+            src_backend_config.device, dst_backend_config.device)
+        self.check_to_device_method_ndarrays()
 
-    def test_to_gpu_from_gpu(self):
-        self.check_to_gpu(cuda.to_gpu(self.x), cuda.to_gpu(self.gx))
+    def test_to_device_method_not_allow_unchain(
+            self, src_backend_config, dst_backend_config):
+        self.setup(src_backend_config)
 
-    @attr.multi_gpu(2)
-    def test_to_gpu_from_another_gpu(self):
-        self.check_to_gpu(cuda.to_gpu(self.x), cuda.to_gpu(self.gx), 1)
+        if self.x_with_graph:
+            with pytest.warns(DeprecationWarning):
+                self.to_device_method(dst_backend_config.device, False)
+        else:
+            # Check that nothing is warned.
+            self.to_device_method(dst_backend_config.device, False)
```

Comment:
It seems it does not check as written in the comment?

Reply:
Hm, it checks (do nothing if it's okay, else raise) just not explicitly but by not wrapping it in an expect-warning context. I think this comment makes sense when you look at his this method as a whole (in particular https://github.com/chainer/chainer/pull/7507/files/757e1b344e5222d43a86377386a247a838887db0#diff-278878af6cd00cabfd7437aa9efff4f2R1062).

Reply:
Is that true? I think warnings do not make tests failure?

Reply:
Yes, they are configured in `setup.cfg` and deprecation warnings in particular are treated as errors https://github.com/chainer/chainer/blob/master/setup.cfg#L12.

Reply:
Ah, I see.
I think "nothing is warned" is misleading: only specific warning types are checked.
By the way, `setup.cfg` is not the test itself: it only defines how tests are run. It's better not rely on it this way.

You can use `pytest.warns`: https://stackoverflow.com/a/45671804

Reply:
Fixed to check the warning count to check cases when no unchaining takes place.